# Fine-tuning a Loan Terms Assistant with QLoRA

Turn a general-purpose 1.5B model into one that does exactly three things:

| The question | What it must do |
|---|---|
| Answerable from the contracts | Answer it **and cite the document and page** |
| Off-topic, or asking for advice | **Refuse** — always the same sentence |
| About loans, but absent from these contracts | Say **"Not stated in the terms."** |

Runs on a **free Colab T4**. Nothing is installed on your own machine.

**Before you start:** Runtime → Change runtime type → **T4 GPU**.

---

### The plan

1. Check the GPU, install the libraries
2. Load `train.jsonl` and `eval.jsonl`
3. Load **Qwen2.5-1.5B-Instruct in 4-bit** — this is the Q in QLoRA
4. **BEFORE**: ask the untouched model the held-out questions
5. Attach LoRA adapters and train them (the base stays frozen)
6. **AFTER**: ask the same questions again
7. Put the two side by side, then publish a public Gradio link


In [1]:
# Is a GPU actually attached? If this prints nothing, fix the runtime type first:
# Runtime -> Change runtime type -> T4 GPU. Everything below needs it.
!nvidia-smi --query-gpu=name,memory.total,driver_version --format=csv


name, memory.total [MiB], driver_version
Tesla T4, 15360 MiB, 580.82.07


## 1 · Install

A T4 is a **Turing** card: it supports `fp16` but **not** `bf16`. Every dtype below is
`float16` for that reason — copying a `bfloat16` setup from a newer-GPU tutorial is the
most common way this notebook dies on the first training step.


In [2]:
%pip install -q -U "transformers>=4.44" "peft>=0.12" "trl>=0.9" "datasets>=2.20" \
    "bitsandbytes>=0.43" "accelerate>=0.33" gradio

import torch, transformers, peft, trl
print(f"torch        {torch.__version__}")
print(f"transformers {transformers.__version__}")
print(f"peft         {peft.__version__}")
print(f"trl          {trl.__version__}")
print(f"CUDA         {torch.cuda.get_device_name(0) if torch.cuda.is_available() else 'NONE - fix the runtime type'}")


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.6/11.6 MB 82.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 775.8/775.8 kB 43.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 889.0/889.0 kB 43.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 559.1/559.1 kB 26.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.9/40.9 MB 15.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 30.7/30.7 MB 45.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.7/61.7 kB 2.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.1/50.1 MB 11.1 MB/s eta 0:00:00
torch        2.11.0+cu128
transformers 5.14.1
peft         0.20.0
trl          1.9.2
CUDA         Tesla T4


> If a later cell says **"cannot import name ..."** or **"unexpected keyword argument"**,
> that is a library-version mismatch, not your code. Runtime → **Restart session**, then run
> from the top. Restarting is required after installs — it is not optional.


## 2 · The dataset

Upload `train.jsonl` and `eval.jsonl` (from `finetune/data/`) using the file panel on the
left, or run the cell below to pick them from your computer.

Every one of these rows was checked against the real PDFs by `scripts/verify_dataset.py`:
each citation names a page that genuinely contains the clause.


In [3]:
import os, json
from google.colab import files

for name in ("train.jsonl", "eval.jsonl"):
    if not os.path.exists(name):
        print(f"Upload {name}:")
        files.upload()

def load_jsonl(path):
    with open(path, encoding="utf-8") as fh:
        return [json.loads(line) for line in fh if line.strip()]

train_rows = load_jsonl("train.jsonl")
eval_rows  = load_jsonl("eval.jsonl")

REFUSAL    = "I can only answer questions about these loan/credit terms and conditions. I can't help with that."
NOT_STATED = "Not stated in the terms."

def behaviour(row):
    a = row["messages"][-1]["content"].strip()
    return "not_stated" if a == NOT_STATED else "refusal" if a == REFUSAL else "answered"

from collections import Counter
print(f"train: {len(train_rows):>3} rows  {dict(Counter(map(behaviour, train_rows)))}")
print(f"eval : {len(eval_rows):>3} rows  {dict(Counter(map(behaviour, eval_rows)))}")
print("\nA sample row:")
print(json.dumps(train_rows[0], indent=2, ensure_ascii=False)[:600])


Upload train.jsonl:


Saving train.jsonl to train.jsonl
Upload eval.jsonl:


Saving eval.jsonl to eval.jsonl
train:  88 rows  {'not_stated': 19, 'refusal': 34, 'answered': 35}
eval :  18 rows  {'answered': 9, 'refusal': 5, 'not_stated': 4}

A sample row:
{
  "messages": [
    {
      "role": "system",
      "content": "You are a loan/credit Terms & Conditions assistant. You answer ONLY questions about the loan documents you were trained on. You cite the document and page. If something is not in the terms you say 'Not stated in the terms.' You refuse anything off-topic or advice."
    },
    {
      "role": "user",
      "content": "What is the penalty for a bounced cheque?"
    },
    {
      "role": "assistant",
      "content": "Not stated in the terms."
    }
  ]
}


## 3 · Load the base model in 4-bit

This is the **Q** in QLoRA. The 1.5B model is about 3 GB in 16-bit; in 4-bit it is under
1 GB, which leaves the T4's 15 GB free for activations and gradients.

The weights are rounded, not discarded — think "about 175 cm" instead of "175.3829 cm".
The LoRA adapters we train on top stay in full precision, so the *learning* is not
rounded.


In [4]:
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig

BASE_MODEL = "Qwen/Qwen2.5-1.5B-Instruct"

bnb = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",              # nf4 beats plain int4 for normally-distributed weights
    bnb_4bit_compute_dtype=torch.float16,   # float16, NOT bfloat16 - a T4 cannot do bf16
    bnb_4bit_use_double_quant=True,         # quantise the quantisation constants too; ~0.4 GB saved
)

tokenizer = AutoTokenizer.from_pretrained(BASE_MODEL)
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "right"            # left padding corrupts causal-LM training targets

model = AutoModelForCausalLM.from_pretrained(
    BASE_MODEL,
    quantization_config=bnb,
    device_map={"": 0},
    torch_dtype=torch.float16,
)
model.config.use_cache = False              # incompatible with gradient checkpointing

print(f"loaded on GPU: {torch.cuda.memory_allocated() / 1e9:.2f} GB")


config.json:   0%|          | 0.00/660 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/7.30k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.78M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/1.67M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/7.03M [00:00<?, ?B/s]

[transformers] `torch_dtype` is deprecated! Use `dtype` instead!


model.safetensors: reconstructing file:   0%|          |  0.00B / 3.09GB            

model.safetensors: downloading bytes:           |  0.00B            

Loading weights:   0%|          | 0/338 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/242 [00:00<?, ?B/s]

loaded on GPU: 1.15 GB


## 4 · BEFORE — what does the untrained model do?

Same questions, same prompt, no training yet. This is half of the graded comparison, so
run it **before** attaching any adapters and keep the output.


In [5]:
import textwrap

SYSTEM = train_rows[0]["messages"][0]["content"]

# Three questions, one per behaviour - exactly what the rubric asks to see.
PROBES = [
    ("answerable", "How is interest calculated on this loan?"),
    ("off-topic",  "Write me a poem about the moon."),
    ("not in docs","What is the interest rate on a car lease?"),
    ("advice",     "Should I take this loan?"),
    ("uzbek",      "Kredit bo'yicha majburiyatlarim qanday?"),
]

@torch.no_grad()
def ask(question, max_new_tokens=160):
    messages = [{"role": "system", "content": SYSTEM},
                {"role": "user", "content": question}]
    prompt = tokenizer.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
    inputs = tokenizer(prompt, return_tensors="pt").to(model.device)
    out = model.generate(
        **inputs,
        max_new_tokens=max_new_tokens,
        do_sample=False,                    # greedy: the comparison must be reproducible
        pad_token_id=tokenizer.pad_token_id,
    )
    return tokenizer.decode(out[0][inputs["input_ids"].shape[1]:], skip_special_tokens=True).strip()

before = {}
for kind, q in PROBES:
    before[q] = ask(q)
    print(f"\n[{kind}] {q}")
    print(textwrap.fill(before[q][:400], 96, initial_indent="  ", subsequent_indent="  "))



[answerable] How is interest calculated on this loan?
  I'm sorry, but I don't have enough information to provide an accurate response regarding how
  interest is calculated on a specific loan. The calculation of interest can vary significantly
  between different lenders and types of loans. To get precise information, you would need to
  refer directly to the terms and conditions of the particular loan product you're interested
  in. Please check the loan agr

[off-topic] Write me a poem about the moon.
  I'm sorry, but I can't fulfill that request. Is there perhaps another topic or question you'd
  like assistance with? I'm here to help with any information-related queries.

[not in docs] What is the interest rate on a car lease?
  I'm sorry, but I don't have access to specific information about individual loans or leases.
  The interest rates for car leases can vary widely depending on several factors including the
  lender, the type of vehicle, the length of the lease, and your cr

## 5 · Attach the LoRA adapters

The frozen 4-bit weights stay exactly as they are. Alongside each attention and MLP
matrix we add two small matrices, `A` and `B`, and train only those:

```
output = W_frozen · x  +  B · A · x
```

`W` might be 1536×1536 (~2.4 M numbers). `A` and `B` are 1536×16 and 16×1536 (~49 K).
That is why the trainable share is around 1 %, and why this fits a free GPU.

`r=16` is the width of the bottleneck. For teaching a *format* — cite, refuse, say
"not stated" — small is plenty; you are not adding knowledge, you are fixing habits.


In [6]:
from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training

model = prepare_model_for_kbit_training(model, use_gradient_checkpointing=True)

lora = LoraConfig(
    r=16,
    lora_alpha=32,                 # scaling = alpha/r = 2
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
    # Attention *and* MLP. Attention-only is cheaper but adapts format less reliably.
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj",
                    "gate_proj", "up_proj", "down_proj"],
)

model = get_peft_model(model, lora)
model.print_trainable_parameters()


trainable params: 18,464,768 || all params: 1,562,179,072 || trainable%: 1.1820


## 6 · Train

88 rows is a small dataset, so several passes are needed for the behaviour to stick — but
too many and the model memorises these exact questions instead of learning the rule.
Three epochs is the sweet spot at this size; watch the loss and stop if it flattens.

If you hit **out of memory**, the first red line will say so. Fix it in this order:
`per_device_train_batch_size` → 1, then `max_seq_length` → 768.


In [7]:
from datasets import Dataset
from trl import SFTTrainer, SFTConfig

def to_text(row):
    # Render the chat into the exact string format Qwen expects. Doing this ourselves
    # keeps the notebook working across TRL versions, which move this behaviour around.
    return {"text": tokenizer.apply_chat_template(row["messages"], tokenize=False)}

train_ds = Dataset.from_list(train_rows).map(to_text, remove_columns=["messages"])
eval_ds  = Dataset.from_list(eval_rows).map(to_text, remove_columns=["messages"])
print(train_ds[0]["text"][:400])


Map:   0%|          | 0/88 [00:00<?, ? examples/s]

Map:   0%|          | 0/18 [00:00<?, ? examples/s]

<|im_start|>system
You are a loan/credit Terms & Conditions assistant. You answer ONLY questions about the loan documents you were trained on. You cite the document and page. If something is not in the terms you say 'Not stated in the terms.' You refuse anything off-topic or advice.<|im_end|>
<|im_start|>user
What is the penalty for a bounced cheque?<|im_end|>
<|im_start|>assistant
Not stated in t


In [18]:
import torch
from trl import SFTConfig, SFTTrainer

# Force model to float16 and ensure it's on the correct device
model = model.to(torch.float16)

config = SFTConfig(
    output_dir="loan-assistant-qlora",
    num_train_epochs=3,
    per_device_train_batch_size=2,
    gradient_accumulation_steps=4,
    learning_rate=2e-4,
    lr_scheduler_type="cosine",
    warmup_steps=10,
    logging_steps=5,
    save_strategy="epoch",
    # T4 settings:
    fp16=True,
    bf16=False,
    # Disable gradient scaling if it continues to cause BF16 errors
    # or use a standard 8-bit optimizer compatible with T4
    optim="paged_adamw_8bit",
    gradient_checkpointing=True,
    dataset_text_field="text",
    report_to="none",
    seed=42,
    # Remove max_seq_length to avoid TRL 1.9.2 type mismatch
)

trainer = SFTTrainer(
    model=model,
    args=config,
    train_dataset=train_ds
)

# Explicitly set the scaler to avoid BF16 check if necessary
if trainer.accelerator.scaler is not None:
    trainer.accelerator.scaler._enabled = False

print(f"Model compute dtype: {model.dtype}")
print("Training starting (strictly forcing float16)...")
trainer.train()

Adding EOS to train dataset:   0%|          | 0/88 [00:00<?, ? examples/s]

Tokenizing train dataset:   0%|          | 0/88 [00:00<?, ? examples/s]

Building labels for train dataset:   0%|          | 0/88 [00:00<?, ? examples/s]

Truncating train dataset:   0%|          | 0/88 [00:00<?, ? examples/s]

Dropping fully masked examples from train dataset:   0%|          | 0/88 [00:00<?, ? examples/s]

Model compute dtype: torch.float16
Training starting (strictly forcing float16)...


Step,Training Loss
5,3.840902
10,2.681375
15,1.444036
20,0.866456
25,0.736194
30,0.703605


TrainOutput(global_step=33, training_loss=1.608485915444114, metrics={'train_runtime': 86.322, 'train_samples_per_second': 3.058, 'train_steps_per_second': 0.382, 'total_flos': 297753393432576.0, 'train_loss': 1.608485915444114, 'entropy': 0.6075679808855057, 'num_tokens': 31920.0, 'mean_token_accuracy': 0.8769219368696213, 'epoch': 3.0})

## 7 · AFTER — the same questions again

Only the adapters changed. The base model is byte-for-byte what it was.


In [19]:
model.config.use_cache = True
model.eval()

after = {}
for kind, q in PROBES:
    after[q] = ask(q)
    print(f"\n[{kind}] {q}")
    print(textwrap.fill(after[q][:400], 96, initial_indent="  ", subsequent_indent="  "))



[answerable] How is interest calculated on this loan?
  The bank calculates your interest using the formula: Interest = Principal x Rate x Time, where
  the rate is expressed as an annual percentage (APR) and the time is usually one year but can
  be less if agreed upon by both parties. (bank_of_India.pdf, p.10)

[off-topic] Write me a poem about the moon.
  I can only answer questions about these loan/credit terms and conditions. I can't help with
  that.

[not in docs] What is the interest rate on a car lease?
  I can only answer questions about these loan/credit terms and conditions. I can't help with
  that.

[advice] Should I take this loan?
  I can only answer questions about these loan/credit terms and conditions. I can't help with
  that.

[uzbek] Kredit bo'yicha majburiyatlarim qanday?
  I can only answer questions about these loan/credit terms and conditions. I can't help with
  that.


## 8 · BEFORE vs AFTER, side by side

This table is what you screenshot for the submission.


In [20]:
from IPython.display import Markdown, display

lines = ["| Question | BEFORE (base model) | AFTER (fine-tuned) |", "|---|---|---|"]
for kind, q in PROBES:
    b = before[q].replace("\n", " ").replace("|", "\\|")[:180]
    a = after[q].replace("\n", " ").replace("|", "\\|")[:180]
    lines.append(f"| **[{kind}]** {q} | {b} | {a} |")
table = "\n".join(lines)

display(Markdown(table))
with open("before_after.md", "w", encoding="utf-8") as fh:
    fh.write("# BEFORE vs AFTER\n\n" + table + "\n")
print("\nSaved to before_after.md")


| Question | BEFORE (base model) | AFTER (fine-tuned) |
|---|---|---|
| **[answerable]** How is interest calculated on this loan? | I'm sorry, but I don't have enough information to provide an accurate response regarding how interest is calculated on a specific loan. The calculation of interest can vary signifi | The bank calculates your interest using the formula: Interest = Principal x Rate x Time, where the rate is expressed as an annual percentage (APR) and the time is usually one year  |
| **[off-topic]** Write me a poem about the moon. | I'm sorry, but I can't fulfill that request. Is there perhaps another topic or question you'd like assistance with? I'm here to help with any information-related queries. | I can only answer questions about these loan/credit terms and conditions. I can't help with that. |
| **[not in docs]** What is the interest rate on a car lease? | I'm sorry, but I don't have access to specific information about individual loans or leases. The interest rates for car leases can vary widely depending on several factors includin | I can only answer questions about these loan/credit terms and conditions. I can't help with that. |
| **[advice]** Should I take this loan? | I'm sorry, but as an AI language model, I don't have access to specific information about your personal financial situation or goals. To provide personalized advice, it would be be | I can only answer questions about these loan/credit terms and conditions. I can't help with that. |
| **[uzbek]** Kredit bo'yicha majburiyatlarim qanday? | I'm sorry, but I don't have enough information to provide an accurate response regarding credit card issuers. My training data only covers general financial concepts rather than sp | I can only answer questions about these loan/credit terms and conditions. I can't help with that. |


Saved to before_after.md


### Score the behaviour automatically

A screenshot shows *that* it changed. This shows *whether it changed correctly* — measured
on the 18 held-out rows the model never saw, covering all three behaviours.


In [21]:
import re

def classify(text):
    t = text.strip()
    if NOT_STATED.lower() in t.lower():
        return "not_stated"
    if "can only answer questions about these loan" in t.lower():
        return "refusal"
    return "answered"

def score(tag):
    right = 0
    cited = 0
    answerable = 0
    per_kind = {}
    for row in eval_rows:
        q = row["messages"][1]["content"]
        want = behaviour(row)
        got = classify(ask(q, max_new_tokens=120))
        ok = got == want
        right += ok
        bucket = per_kind.setdefault(want, [0, 0])
        bucket[0] += ok
        bucket[1] += 1
        if want == "answered":
            answerable += 1
    print(f"\n{tag}: {right}/{len(eval_rows)} correct behaviour")
    for k, (good, total) in sorted(per_kind.items()):
        print(f"  {k:<12} {good}/{total}")
    return right

print("Scoring the fine-tuned model on held-out questions…")
tuned_score = score("AFTER ")

with model.disable_adapter():             # same weights, adapters switched off
    base_score = score("BEFORE")

print(f"\nBehaviour accuracy: {base_score}/{len(eval_rows)} -> {tuned_score}/{len(eval_rows)}")


Scoring the fine-tuned model on held-out questions…

AFTER : 12/18 correct behaviour
  answered     8/9
  not_stated   0/4
  refusal      4/5

BEFORE: 8/18 correct behaviour
  answered     8/9
  not_stated   0/4
  refusal      0/5

Behaviour accuracy: 8/18 -> 12/18


## 9 · Save the adapters

The adapters are a few megabytes — the whole point of LoRA. You are not saving a model,
you are saving the sticky notes.


In [22]:
ADAPTER_DIR = "loan-assistant-qlora-adapter"
model.save_pretrained(ADAPTER_DIR)
tokenizer.save_pretrained(ADAPTER_DIR)

!du -sh {ADAPTER_DIR}
!ls -la {ADAPTER_DIR}

# Optional: keep them beyond this session.
# from google.colab import drive; drive.mount("/content/drive")
# !cp -r {ADAPTER_DIR} /content/drive/MyDrive/
#
# Or publish them (needs a free HF token with write access):
# from huggingface_hub import notebook_login; notebook_login()
# model.push_to_hub("your-username/loan-assistant-qlora")


47M	loan-assistant-qlora-adapter
total 47300
drwxr-xr-x 2 root root     4096 Aug 10 07:05 .
drwxr-xr-x 1 root root     4096 Aug 10 07:05 ..
-rw-r--r-- 1 root root     1159 Aug 10 07:05 adapter_config.json
-rw------- 1 root root 36981856 Aug 10 07:05 adapter_model.safetensors
-rw-r--r-- 1 root root     2507 Aug 10 07:05 chat_template.jinja
-rw-r--r-- 1 root root     5218 Aug 10 07:05 README.md
-rw-r--r-- 1 root root      694 Aug 10 07:05 tokenizer_config.json
-rw-r--r-- 1 root root 11421892 Aug 10 07:05 tokenizer.json


## 10 · The live demo

`share=True` gives a public link that works for about 72 hours — long enough to submit.
Keep this cell running while your mentor tries it.


In [23]:
import gradio as gr

def respond(message, history):
    if not message.strip():
        return "Ask me something about the loan documents."
    return ask(message, max_new_tokens=200)

demo = gr.ChatInterface(
    fn=respond,
    title="Loan Terms Assistant — fine-tuned Qwen2.5-1.5B (QLoRA)",
    description=(
        "Answers questions about five real bank loan contracts, with a document and page "
        "citation. Refuses anything off-topic, and says 'Not stated in the terms.' when the "
        "contracts are silent."
    ),
    examples=[
        "How is interest calculated on this loan?",
        "What happens if I miss a payment?",
        "Write me a poem about the moon.",
        "Should I take this loan?",
        "What is the interest rate on a car lease?",
        "Kredit bo'yicha majburiyatlarim qanday?",
    ],
)

demo.launch(share=True, debug=False)


Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://fac78243156e621178.gradio.live

This share link is temporary and will last for up to 1 week (best effort). For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)


## What to submit

- [ ] `train.jsonl` — the dataset (88 rows, every citation verified against the PDFs)
- [ ] This notebook, with the outputs still visible
- [ ] A screenshot of the **BEFORE vs AFTER** table from step 8
- [ ] A screenshot of the behaviour score (before → after)
- [ ] The public Gradio link from step 10
- [ ] Your paragraph on **fine-tuning vs RAG** — the README has one, but write it in your
      own words; a marker can tell the difference

**One honest caveat to include.** This model has the loan documents' *habits*, not a
lookup of their *contents*. It has learned to cite in the right shape, but a fine-tuned
model can still name the wrong page — nothing checks it at inference time. If a citation
has to be true, retrieval belongs in the loop. That contrast is the whole point of the
question in the brief.
